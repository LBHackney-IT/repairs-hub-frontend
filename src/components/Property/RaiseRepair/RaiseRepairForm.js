import PropTypes from 'prop-types'
import { useState } from 'react'
import { useForm } from 'react-hook-form'
import TenureDetails from '../TenureDetails'
import BackButton from '../../Layout/BackButton/BackButton'
import {
  Select,
  PrimarySubmitButton,
  CharacterCountLimitedTextArea,
  TextInput,
} from '../../Form'
import TradeContractorRateScheduleItemView from './TradeContractorRateScheduleItemView'
import Contacts from '../Contacts/Contacts'
import WarningText from '../../Template/WarningText'
import { buildScheduleRepairFormData } from '../../../utils/hact/schedule-repair/raise-repair-form'

const RaiseRepairForm = ({
  propertyReference,
  address,
  hierarchyType,
  canRaiseRepair,
  locationAlerts,
  personAlerts,
  tenure,
  priorities,
  trades,
  contacts,
  onFormSubmit,
  raiseLimit,
}) => {
  const { register, handleSubmit, errors } = useForm()
  const [priorityCode, setPriorityCode] = useState('')
  const priorityList = priorities.map((priority) => priority.description)

  const [totalCost, setTotalCost] = useState('')
  const overSpendLimit = totalCost > raiseLimit

  const onSubmit = async (formData) => {
    const scheduleRepairFormData = buildScheduleRepairFormData(formData)

    onFormSubmit(scheduleRepairFormData)
  }

  const getPriorityObjectByDescription = (description) => {
    return priorities.filter(
      (priority) => priority.description == description
    )[0]
  }
  const getPriorityObjectByCode = (code) => {
    return priorities.filter((priority) => priority.priorityCode == code)[0]
  }

  const onPrioritySelect = (event) => {
    const priorityObject = getPriorityObjectByDescription(event.target.value)

    setPriorityCode(priorityObject?.priorityCode)
  }

  const updatePriority = (
    description,
    code,
    rateScheduleItemsLength,
    existingHigherPriorityCode
  ) => {
    if (existingHigherPriorityCode) {
      description = getPriorityObjectByCode(existingHigherPriorityCode)
        ?.description
    }

    if (priorityList.includes(description)) {
      const existingCode = parseInt(
        document.getElementById('priorityCode').value
      )
      // Update priority when SOR code has priority attached if:
      // Priority description is blank, or there's only one sor code entry, or
      // when removing an SOR there's an existing entry with higher priority, or
      // the selected priority code is less than existing priority codes
      // (Higher priority as code gets lower)
      if (
        !existingCode ||
        rateScheduleItemsLength <= 1 ||
        existingHigherPriorityCode ||
        code < existingCode
      ) {
        if (errors?.priorityDescription) {
          delete errors.priorityDescription
        }

        document.getElementById('priorityDescription').value = description

        setPriorityCode(
          getPriorityObjectByDescription(description)?.priorityCode
        )
      }
    } else {
      console.error(
        `Priority: "${description}" is not included in the available priorities list`
      )
    }
  }

  return (
    <>
      <BackButton />
      <div className="govuk-grid-row">
        <div className="govuk-grid-column-two-thirds">
          <span className="govuk-caption-l lbh-caption">New repair</span>
          <h1 className="lbh-heading-h1 govuk-!-margin-bottom-2">
            {hierarchyType.subTypeDescription}: {address.addressLine}
          </h1>

          <div className="lbh-body-s">
            <TenureDetails
              canRaiseRepair={canRaiseRepair}
              tenure={tenure}
              locationAlerts={locationAlerts}
              personAlerts={personAlerts}
            />
          </div>

          <h2 className="lbh-heading-h2 govuk-!-margin-top-6">
            Repair task details
          </h2>

          <form
            role="form"
            id="repair-request-form"
            onSubmit={handleSubmit(onSubmit)}
          >
            <TradeContractorRateScheduleItemView
              register={register}
              errors={errors}
              trades={trades}
              propertyReference={propertyReference}
              register={register}
              errors={errors}
              isContractorUpdatePage={false}
              updatePriority={updatePriority}
              getPriorityObjectByCode={getPriorityObjectByCode}
              setTotalCost={setTotalCost}
            />
            <Select
              name="priorityDescription"
              label="Task priority"
              options={priorityList}
              onChange={onPrioritySelect}
              disabled={true}
              required={true}
              register={register({
                required: 'Please select a priority',
                validate: (value) =>
                  priorityList.includes(value) || 'Priority is not valid',
              })}
              error={errors && errors.priorityDescription}
              widthClass="govuk-!-width-full"
            />
            <input
              id="priorityCode"
              name="priorityCode"
              label="priorityCode"
              type="hidden"
              value={priorityCode}
              ref={register}
            />
            <CharacterCountLimitedTextArea
              name="descriptionOfWork"
              label="Repair description"
              required={true}
              maxLength={230}
              requiredText="Please enter a repair description"
              register={register}
              error={errors && errors.descriptionOfWork}
            />
            <input
              id="propertyReference"
              name="propertyReference"
              label="propertyReference"
              type="hidden"
              value={propertyReference}
              ref={register}
            />
            <input
              id="shortAddress"
              name="shortAddress"
              label="shortAddress"
              type="hidden"
              value={address.shortAddress}
              ref={register}
            />
            <input
              id="postalCode"
              name="postalCode"
              label="postalCode"
              type="hidden"
              value={address.postalCode}
              ref={register}
            />
            <Contacts contacts={contacts} />
            <TextInput
              name="callerName"
              label="Caller name"
              required={true}
              register={register({
                required: 'Please add caller name',
                maxLength: {
                  value: 50,
                  message:
                    'You have exceeded the maximum amount of 50 characters',
                },
              })}
              error={errors && errors.callerName}
            />
            <TextInput
              name="contactNumber"
              label="Contact number"
              required={true}
              register={register({
                required: 'Please add contact number',
              })}
              error={errors && errors.contactNumber}
            />

            {overSpendLimit && (
              <WarningText text="The works order cost exceeds the approved spending limit and will be sent to a manager for authorisation" />
            )}

            <PrimarySubmitButton label="Create works order" />
          </form>
        </div>
      </div>
    </>
  )
}

RaiseRepairForm.propTypes = {
  propertyReference: PropTypes.string.isRequired,
  address: PropTypes.object.isRequired,
  hierarchyType: PropTypes.object.isRequired,
  canRaiseRepair: PropTypes.bool.isRequired,
  locationAlerts: PropTypes.array.isRequired,
  personAlerts: PropTypes.array.isRequired,
  tenure: PropTypes.object,
  trades: PropTypes.array.isRequired,
  priorities: PropTypes.array.isRequired,
  onFormSubmit: PropTypes.func.isRequired,
}

export default RaiseRepairForm
