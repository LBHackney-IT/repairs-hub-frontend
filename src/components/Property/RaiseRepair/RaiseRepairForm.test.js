import { render } from '@testing-library/react'
import RaiseRepairForm from './RaiseRepairForm'

describe('RaiseRepairForm component', () => {
  const props = {
    property: {
      propertyReference: '00012345',
      address: {
        shortAddress: '16 Pitcairn House  St Thomass Square',
        postalCode: 'E9 6PT',
        addressLine: '16 Pitcairn House',
        streetSuffix: 'St Thomass Square',
      },
      hierarchyType: {
        levelCode: '7',
        subTypeCode: 'DWE',
        subTypeDescription: 'Dwelling',
      },
    },
    alerts: {
      locationAlert: [
        {
          alertCode: 'DIS',
          description: 'Property Under Disrepair',
          startDate: '2011-02-16',
          endDate: null,
        },
      ],
      personAlert: [
        {
          alertCode: 'DIS',
          description: 'Property Under Disrepair',
          startDate: '2011-08-16',
          endDate: null,
        },
      ],
    },
    tenure: {
      typeCode: 'SEC',
      typeDescription: 'Secure',
      canRaiseRepair: true,
    },
    sorCodes: [
      {
        customCode: 'DES5R003',
        customName: 'Immediate call outs',
        priority: {
          priorityCode: 1,
          description: 'I - Immediate (2 hours)',
        },
        sorContractor: {
          reference: 'H01',
        },
      },
      {
        customCode: 'DES5R004',
        customName: 'Emergency call outs',
        priority: {
          priorityCode: 2,
          description: 'E - Emergency (24 hours)',
        },
        sorContractor: {
          reference: 'H01',
        },
      },
    ],
    onFormSubmit: jest.fn(),
  }

  it('should render properly', () => {
    const { asFragment } = render(
      <RaiseRepairForm
        address={props.property.address}
        hierarchyType={props.property.hierarchyType}
        tenure={props.tenure}
        locationAlerts={props.alerts.locationAlert}
        personAlerts={props.alerts.personAlert}
        sorCodes={props.sorCodes}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
