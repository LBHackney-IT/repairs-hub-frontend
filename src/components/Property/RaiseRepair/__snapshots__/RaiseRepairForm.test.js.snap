// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RaiseRepairForm component should render properly 1`] = `
<DocumentFragment>
  <a
    class="govuk-back-link lbh-back-link"
    role="button"
  >
    Back
  </a>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds"
    >
      <span
        class="govuk-caption-l"
      >
        New repair
      </span>
      <h1
        class="lbh-heading-l govuk-!-margin-bottom-2"
      >
        Dwelling: 16 Pitcairn House
      </h1>
      <div
        class="govuk-body-s"
      >
        <ul
          class="hackney-property-alerts"
        >
          <li
            class="bg-turquoise"
          >
            Tenure: Secure
          </li>
          <li
            class="bg-orange"
          >
            Address Alert:  (
            <strong />
            )
          </li>
          <li
            class="bg-orange"
          >
            Contact Alert:  (
            <strong />
            )
          </li>
        </ul>
      </div>
      <h2
        class="lbh-heading-h2 govuk-!-margin-top-6"
      >
        Repair task details
      </h2>
      <form
        id="repair-request-form"
        role="form"
      >
        <div
          class="govuk-form-group lbh-form-group"
          id="trade-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="trade"
          >
            Trade 
            <span
              class="govuk-required"
            >
              *
            </span>
          </label>
          <input
            autocomplete="off"
            class="govuk-select lbh-select govuk-!-width-one-half"
            data-testid="trade"
            id="trade"
            list="autocomplete-list-trade"
            name="trade"
          />
          <datalist
            id="autocomplete-list-trade"
          >
            <option
              value="DOOR ENTRY ENGINEER - DE"
            />
            <option
              value="Plumbing - PL"
            />
          </datalist>
        </div>
        <input
          id="tradeCode"
          name="tradeCode"
          type="hidden"
          value=""
        />
        <div
          class="min-height-100"
        >
          <div
            class="govuk-form-group lbh-form-group"
            id="contractor-form-group"
          >
            <label
              class="govuk-label lbh-label"
              for="contractor"
            >
              Contractor 
              <span
                class="govuk-required"
              >
                *
              </span>
            </label>
            <select
              class="govuk-select lbh-select govuk-!-width-one-half"
              data-testid="contractor"
              disabled=""
              id="contractor"
              name="contractor"
            >
              <option
                value=""
              />
            </select>
          </div>
        </div>
        <input
          id="contractorRef"
          name="contractorRef"
          type="hidden"
          value=""
        />
        <div
          class="min-height-120 govuk-!-margin-bottom-6"
        >
          <div>
            <div
              class="govuk-grid-row rate-schedule-item display-flex align-items-center position-relative"
              id="rate-schedule-item-0"
            >
              <div
                class="govuk-grid-column-two-thirds"
              >
                <div
                  class="govuk-form-group lbh-form-group"
                  id="rateScheduleItems[0][code]-form-group"
                >
                  <label
                    class="govuk-label lbh-label"
                    for="rateScheduleItems[0][code]"
                  >
                    SOR Code 
                    <span
                      class="govuk-required"
                    >
                      *
                    </span>
                  </label>
                  <select
                    class="govuk-select lbh-select govuk-!-width-full"
                    data-testid="rateScheduleItems[0][code]"
                    disabled=""
                    id="rateScheduleItems[0][code]"
                    name="rateScheduleItems[0][code]"
                  >
                    <option
                      selected=""
                      value=""
                    />
                  </select>
                </div>
                <input
                  id="rateScheduleItems[0][description]"
                  name="rateScheduleItems[0][description]"
                  type="hidden"
                />
                <input
                  id="rateScheduleItems[0][cost]"
                  name="rateScheduleItems[0][cost]"
                  type="hidden"
                  value=""
                />
              </div>
              <div
                class="govuk-grid-column-one-third"
              >
                <div
                  class="govuk-form-group lbh-form-group"
                  id="rateScheduleItems[0][quantity]-form-group"
                >
                  <label
                    class="govuk-label lbh-label"
                    for="rateScheduleItems[0][quantity]"
                  >
                    Quantity 
                    <span
                      class="govuk-required"
                    >
                      *
                    </span>
                  </label>
                  <input
                    class="govuk-input lbh-input govuk-!-width-full"
                    disabled=""
                    id="rateScheduleItems[0][quantity]"
                    name="rateScheduleItems[0][quantity]"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <a
              class="repairs-hub-link govuk-body-s"
              href="#"
            >
              + Add another SOR code
            </a>
          </div>
        </div>
        <div
          class="govuk-form-group lbh-form-group"
          id="priorityDescription-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="priorityDescription"
          >
            Task priority 
            <span
              class="govuk-required"
            >
              *
            </span>
          </label>
          <select
            class="govuk-select lbh-select govuk-!-width-full"
            data-testid="priorityDescription"
            disabled=""
            id="priorityDescription"
            name="priorityDescription"
          >
            <option
              value=""
            />
            <option
              value="1 [I] IMMEDIATE"
            >
              1 [I] IMMEDIATE
            </option>
            <option
              value="2 [E] EMERGENCY"
            >
              2 [E] EMERGENCY
            </option>
            <option
              value="4 [U] URGENT"
            >
              4 [U] URGENT
            </option>
            <option
              value="5 [N] NORMAL"
            >
              5 [N] NORMAL
            </option>
          </select>
        </div>
        <input
          id="priorityCode"
          label="priorityCode"
          name="priorityCode"
          type="hidden"
          value=""
        />
        <div
          class="govuk-form-group lbh-form-group"
          id="descriptionOfWork-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="descriptionOfWork"
          >
            Repair description 
            <span
              class="govuk-required"
            >
              *
            </span>
          </label>
          <textarea
            aria-describedby="descriptionOfWork-hint descriptionOfWork-error"
            class="govuk-textarea lbh-textarea"
            id="descriptionOfWork"
            name="descriptionOfWork"
            rows="4"
          />
        </div>
        <span
          class="govuk-hint govuk-!-margin-bottom-6"
        >
          You have 250 characters remaining.
        </span>
        <input
          id="propertyReference"
          label="propertyReference"
          name="propertyReference"
          type="hidden"
          value="00012345"
        />
        <input
          id="shortAddress"
          label="shortAddress"
          name="shortAddress"
          type="hidden"
          value="16 Pitcairn House  St Thomass Square"
        />
        <div
          class="govuk-warning-text lbh-warning-text"
        >
          <span
            aria-hidden="true"
            class="govuk-warning-text__icon"
          >
            !
          </span>
          <strong
            class="govuk-warning-text__text"
          >
            <span
              class="govuk-warning-text__assistive"
            >
              Warning
            </span>
            No contact details available for this property
          </strong>
        </div>
        <div
          class="govuk-form-group lbh-form-group"
          id="callerName-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="callerName"
          >
            Caller name 
          </label>
          <input
            class="govuk-input lbh-input undefined"
            id="callerName"
            name="callerName"
            type="text"
            value=""
          />
        </div>
        <div
          class="govuk-form-group lbh-form-group"
          id="contactNumber-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="contactNumber"
          >
            Contact number 
          </label>
          <input
            class="govuk-input lbh-input undefined"
            id="contactNumber"
            name="contactNumber"
            type="text"
            value=""
          />
        </div>
        <div
          class="govuk-form-group lbh-form-group"
        >
          <button
            class="govuk-button lbh-button"
            data-module="govuk-button"
            type="submit"
          >
            Create works order
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
